File: Trunk/IAS/CommonFeature/Session/UIH.XR.SessionUtility/SessionUtility.cs
using System.Text;
using UIH.Mcsf.Pipeline.Data;
using UIH.Mcsf.Core;
using UIH.XR.Basis.Proto;

namespace UIH.XR.SessionManager
{
    public class SessionUtility
    {
        /// <summary>
        /// Serialize dataHead to string
        /// </summary>
        /// <param name="dataHead">dataHead object</param>
        /// <returns>serializeDataHead</returns>
        public static string SerializeDataHead(DicomAttributeCollection dataHead)
        {
            byte[] bValue = null;
            bool result = dataHead.Serialize(out bValue);
            return result ? Encoding.Default.GetString(bValue) : null;
        }

        /// <summary>
        /// Deserialize string to dataHead
        /// </summary>
        /// <param name="serializedDataHead">a serialized dataHead</param>
        /// <returns>dataHead object</returns>
        public static DicomAttributeCollection DeserializeDataHead(string serializedDataHead)
        {
            return DicomAttributeCollection.Deserialize(Encoding.Default.GetBytes(serializedDataHead));
        }

        /// <summary>
        /// Create commandContext
        /// </summary>
        /// <param name="receiver">receiver of communication</param>
        /// <param name="commandID">commandID of communication</param>
        /// <param name="serializeObject">message needs to send</param>
        /// <returns>CommandContext object</returns>
        public static CommandContext CreateCommandContext(string receiver, int commandID, byte[] serializeObject)
        {
            CommandContext context = new CommandContext();
            context.sReceiver = receiver;
            context.iCommandId = commandID;
            context.sSerializeObject = serializeObject;
            return context;
        }

        /// <summary>
        /// Create a protobuffer,type is byte[]
        /// </summary>
        /// <param name="commandID">cmd</param>
        /// <param name="parArray">parameter array</param>
        /// <returns>protobuffer</returns>
        public static byte[] RequestCommandProtoBuffer(int commandID, string[] parArray)
        {
            SessionRequestCommand.Builder builder = new SessionRequestCommand.Builder();
            builder.SetCommandId(commandID);
            builder.AddRangeParameters(parArray);
            return builder.Build().ToByteArray();
        }
    }
}

File: Trunk/IAS/CommonFeature/Session/XRSessionUtility/SessionUtility.cpp
// This is the main DLL file.

#include "stdafx.h"
#include "McsfDataHeader/mcsf_data_header_common.h"
#include "SessionUtility.h"

using namespace std;
#pragma comment(lib, "McsfLogger.lib")
#pragma comment(lib, "McsfDataHeader.lib")
MCSF_DATAHEADER_BEGIN_NAMESPACE 

string CSessionUtility::SerializeDataHead(IDataHeaderElementMap *m_pDataHeaderElemMap)
 {
	 string sSerializedOutput("");
	 IDataHeaderElementMap::Serialize(m_pDataHeaderElemMap,&sSerializedOutput);
	 return sSerializedOutput;
 }

shared_ptr<IDataHeaderElementMap> CSessionUtility::DeserializeDataHead(string serializedDataHead)
{
	shared_ptr<IDataHeaderElementMap>
		pHeader(Mcsf::IDataHeaderElementMap::CreateDataHeader());
	IDataHeaderElementMap::Deserialize(serializedDataHead,pHeader.get());
	return pHeader;
}

CommandContext* CSessionUtility::CreateCommandContext(string receiver, int commandID, string serializeObject)
{
	Mcsf::CommandContext* context = new Mcsf::CommandContext();
	context->sReceiver = receiver;
	context->iCommandId = commandID;
	context->sSerializeObject = serializeObject;
	return context;
}

string CSessionUtility::RequestCommandProtoBuffer(int commandID,string parArray[])
{
	UIH::XR::Basis::Proto::SessionRequestCommand sRequestCommand;
	sRequestCommand.set_commandid(commandID);
	for(int i=0;i<sizeof(parArray)/sizeof(string);i++)
	{
		if(parArray[i]=="0")continue;
		sRequestCommand.set_parameters(i,parArray[i]);
	}
	string sRequestCommandString= "";
	sRequestCommand.SerializeToString(&sRequestCommandString);
	return sRequestCommandString;
}

MCSF_DATAHEADER_END_NAMESPACE